src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/Dwarf/DwarfInfoWriter.cs

```diff
@@ -0,0 +1,218 @@
+            }
+        }
+
+        public void Write(ReadOnlySpan<byte> buffer) => _infoSectionWriter.Write(buffer);
+        public void WriteULEB128(ulong value) => _infoSectionWriter.WriteULEB128(value);
+        public void WriteUInt8(byte value) => _infoSectionWriter.WriteLittleEndian<byte>(value);
+        public void WriteUInt16(ushort value) => _infoSectionWriter.WriteLittleEndian<ushort>(value);
+        public void WriteUInt32(uint value) => _infoSectionWriter.WriteLittleEndian<uint>(value);
+        public void WriteUInt64(ulong value) => _infoSectionWriter.WriteLittleEndian<ulong>(value);
+
+        public void WriteAddressSize(ulong value)
+        {
+            switch (TargetPointerSize)
+            {
+                case 4: WriteUInt32((uint)value); break;
+                case 8: WriteUInt64(value); break;
+                default: throw new NotSupportedException();
+            }
+        }
+
+        public void WriteStringReference(string value)
+        {
+            long stringsOffset = _stringTableWriter.Position;
+            _stringTableWriter.WriteUtf8String(value);
+
+            Debug.Assert(stringsOffset < uint.MaxValue);
+            _infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_str", stringsOffset);
+        }
+
+        public void WriteInfoAbsReference(long offset)
+        {
+            Debug.Assert(offset < uint.MaxValue);
+            _infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_info", offset);
+        }
+
+        public void WriteInfoReference(uint typeIndex)
+        {
+            uint offset = _builder.ResolveOffset(typeIndex);
+
+            if (offset == 0)
+            {
+                // Late bound forward reference
+                var data = new byte[sizeof(uint)];
+                _lateBoundReferences.Add(new InfoReference(typeIndex, (int)_infoSectionWriter.Position, data));
+                _infoSectionWriter.EmitData(data);
+            }
+            else
+            {
+                _infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_info", offset);
+            }
+        }
+
+        public void WriteCodeReference(string sectionSymbolName, long offset = 0)
+        {
+            Debug.Assert(offset >= 0);
+            _infoSectionWriter.EmitSymbolReference(_codeRelocType, sectionSymbolName, offset);
+        }
+
+        public void WriteLineReference(long offset)
+        {
+            Debug.Assert(offset < uint.MaxValue);
+            _infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_line", offset);
+        }
+
+        public DwarfExpressionBuilder GetExpressionBuilder()
+        {
+            _expressionBufferWriter.Clear();
+            return new DwarfExpressionBuilder(TargetArchitecture, TargetPointerSize, _expressionBufferWriter);
+        }
+
+        public void WriteExpression(DwarfExpressionBuilder expressionBuilder)
+        {
```

Comment:
Since this uses a writer (not a builder), it should be passed a writer.

Reply:
or at least retrieve the writer from the builder.

Reply:
it is strange that the only parameter is unused

Reply:
I agree it looks strange at first but the idea is that you build the expression in memory first (so it’s more like a `StringBuilder`, hence the `-Builder` suffix) and then you write it either as single expression, or as location list with multiple expressions at multiple code locations.

The DWARF format treats these two cases quite differently since the final form of the expression uses different length prefix and ends up in different section of the object file.

The API on `DwarfInfoWriter` is thus shaped to make it easy for the caller to write the expression in one of two ways.

For single expression,
```csharp
                var expressionBuilder = writer.GetExpressionBuilder();
                // Build the expression in expressionBuilder
                writer.WriteExpression(expressionBuilder);
```

For location list,
```csharp
                writer.WriteStartLocationList();
                foreach (...)
                {
                    var expressionBuilder = writer.GetExpressionBuilder();
                    // Build the expression in expressionBuilder
                    writer.WriteLocationListExpression(..., start, end, expressionBuilder);
                }
                writer.WriteEndLocationList();
```

In both of these scenarios you are allowed to only call `GetExpressionBuilder` once to "rent it" and the `WriteExpression`/`WriteLocationListExpression` method to "return it". That's partly an optimization to reuse the same `ArrayBufferWriter` and implementation detail. Ideally we want to move to `PooledByteBufferWriter` (currently a private type in System.Text.Json code). Then we can remove the `_expressionBufferWriter` field and replace it with accessing the buffer writer from the `expressionBuilder` parameter.

*I didn't want to do the refactoring to make `PooledByteBufferWriter` available as part of this PR. I think there's also an API proposal to make it available as public API which would make it easier.*

Reply:
Makes more sense with the explanation.